### ISSUE TITLE
Phase 8.6: Smart Recommendation Features & ML Enhancements

### ISSUE THREAD

**Issue by jpequegn** [jpequegn/ai-coach#134]
## Overview
Implement advanced features including ML-based ranking, collaborative filtering, time-series forecasting, and recommendation protocols.

**Parent Issue**: #118 - Phase 8: Advanced Recovery Recommendation Engine
**Estimated Time**: 5-6 days

## Dependencies
- ✅ #129 - Recommendation Library Foundation
- ✅ #130 - User Recommendation Tracking
- ✅ #131 - Recommendation Selection Engine
- ✅ #132 - User Recovery Profiles
- ✅ #133 - Recommendation Effectiveness Measurement

### ML-Based Recommendation Ranking

- [ ] Implement feature engineering:
  ```rust
  struct RecommendationFeatures {
      // User features
      user_completion_rate: f64,
      user_avg_rating: f64,
      user_activity_level: f64,
      
      // Recommendation features
      template_effectiveness: f64,
      category_preference: f64,
      time_since_last_similar: f64,
      
      // Context features
      recovery_score: f64,
      time_of_day: f64,
      day_of_week: f64,
      season: String,
  }
  ```

- [ ] Train gradient boosted trees model:
  - Target: Predicted completion + effectiveness
  - Features: User history, recommendation metadata, context
  - Training data: Historical completions and outcomes
  - Model: linfa GBT or LightGBM

- [ ] Implement ML-based ranking:
  ```rust
  async fn rank_with_ml(
      recommendations: Vec<RecommendationTemplate>,
      user_id: Uuid,
      context: Context,
  ) -> Result<Vec<ScoredRecommendation>> {
      let features = extract_features(recommendations, user_id, context).await?;
      let predictions = model.predict(features)?;
      
      recommendations
          .into_iter()
          .zip(predictions)
          .map(|(rec, score)| ScoredRecommendation { rec, score })
          .sorted_by(|a, b| b.score.partial_cmp(&a.score).unwrap())
          .collect()
  }
  ```

### Collaborative Filtering

- [ ] Implement user similarity:
  ```rust
  async fn find_similar_users(
      user_id: Uuid,
      limit: usize,
  ) -> Result<Vec<SimilarUser>> {
      // Compare on:
      // - Recovery patterns (HRV, sleep, stress)
      // - Completion patterns (categories, timing)
      // - Effectiveness patterns (what works)
      
      // Return top N similar users
  }
  ```

- [ ] Recommend from similar users:
  ```rust
  async fn collaborative_recommendations(
      user_id: Uuid,
  ) -> Result<Vec<Recommendation>> {
      let similar_users = find_similar_users(user_id, 10).await?;
      
      // Find recommendations that worked well for similar users
      // but this user hasn't tried yet
      
      let successful_recs = get_successful_recommendations(similar_users).await?;
      let user_history = get_user_history(user_id).await?;
      
      successful_recs
          .into_iter()
          .filter(|r| !user_history.contains(r))
          .collect()
  }
  ```

- [ ] Add social proof:
  - "Users like you found this helpful (87% completion)"
  - "Top recommendation for similar athletes"

### Time-Series Forecasting

- [ ] Implement recovery trajectory prediction:
  ```rust
  async fn predict_recovery_trajectory(
      user_id: Uuid,
      days_ahead: i32,
  ) -> Result<Vec<RecoveryPrediction>> {
      // Use historical recovery scores
      // Account for seasonal patterns
      // Consider current trends
      
      // Predict next N days
  }
  ```

- [ ] Proactive intervention recommendations:
  - If trajectory shows declining recovery → suggest preventive actions
  - If peak recovery predicted → suggest optimal training window
  - If stress accumulation forecast → recommend stress management

- [ ] Optimal timing suggestions:
  ```rust
  async fn suggest_optimal_timing(
      user_id: Uuid,
      recommendation: &Recommendation,
  ) -> Result<OptimalTiming> {
      // Analyze user patterns
      // Find when similar recommendations worked best
      // Consider current schedule
      
      OptimalTiming {
          best_time: "18:00-19:00",
          reasoning: "You complete sleep recommendations 2x more when done in evening",
          alternative_times: vec!["morning", "afternoon"],
      }
  }
  ```

### Recovery Protocols

- [ ] Create multi-recommendation protocols:
  ```rust
  struct RecoveryProtocol {
      name: String,
      description: String,
      recommendations: Vec<Uuid>,
      sequence: ProtocolSequence, // parallel or sequential
      duration: Duration,
      effectiveness_score: f64,
  }
  
  enum ProtocolSequence {
      Parallel,    // All at once
      Sequential,  // One after another
      Phased,      // Over multiple days
  }
  ```

- [ ] Pre-built protocols:
  - "Post-Hard-Workout Recovery" (24h protocol)
  - "Sleep Optimization Week" (7-day protocol)
  - "Stress Reset" (3-day protocol)
  - "Travel Recovery" (pre/during/post travel)
  - "Injury Prevention" (ongoing protocol)

- [ ] Protocol API:
  - `GET /api/v1/recovery/protocols`
  - `POST /api/v1/recovery/protocols/activate`
  - `GET /api/v1/recovery/protocols/active`

### Progressive Recommendations

- [ ] Implement complexity progression:
  ```rust
  async fn select_with_progression(
      user_id: Uuid,
      template: &RecommendationTemplate,
  ) -> Result<Recommendation> {
      let user_experience = get_user_experience_level(user_id).await?;
      
      match user_experience {
          ExperienceLevel::Beginner => template.get_simple_version(),
          ExperienceLevel::Intermediate => template.get_standard_version(),
          ExperienceLevel::Advanced => template.get_advanced_version(),
      }
  }
  ```

- [ ] Habit building sequences:
  - Week 1: Simple, easy wins
  - Week 2-3: Add complexity
  - Week 4+: Advanced techniques
  - Track mastery and progression

### Context-Aware Features

- [ ] Weather-based recommendations:
  - Rainy day → Indoor recovery options
  - Sunny day → Outdoor activities
  - Extreme temp → Modified activities

- [ ] Location-based recommendations:
  - At home → Full equipment options
  - At gym → Gym-specific recovery
  - Traveling → Travel-friendly options

- [ ] Schedule-aware recommendations:
  - Busy day → Quick 5-min techniques
  - Free evening → Comprehensive protocols
  - Weekend → Longer recovery sessions

### API Enhancements

- [ ] `POST /api/v1/recovery/recommendations/ml-ranked`
  - ML-based ranking (requires model training)
  - Returns predicted effectiveness scores

- [ ] `GET /api/v1/recovery/recommendations/collaborative`
  - Recommendations from similar users
  - Social proof included

- [ ] `GET /api/v1/recovery/recommendations/proactive`
  - Based on forecasted recovery trajectory
  - Preventive interventions

- [ ] `GET /api/v1/recovery/protocols`
  - Pre-built recovery protocols
  - Customizable based on user needs

## Deliverables

- [ ] ML-based ranking model
- [ ] Collaborative filtering system
- [ ] Recovery trajectory forecasting
- [ ] Recovery protocol framework
- [ ] Progressive recommendation system
- [ ] Context-aware features
- [ ] Enhanced API endpoints

## Acceptance Criteria

- ML model achieves >60% accuracy on completion prediction
- Collaborative filtering suggests relevant alternatives
- Trajectory forecasting provides 3-day outlook
- Recovery protocols combine multiple recommendations effectively
- Progressive system adapts to user experience level
- Context-aware features influence recommendation selection

## Testing

- [ ] Unit tests for ML feature extraction
- [ ] Test collaborative filtering similarity
- [ ] Validate trajectory prediction accuracy
- [ ] Test protocol activation and tracking
- [ ] Integration tests for all new endpoints

## ML Model Training

```bash
# Train ranking model
cargo run --bin train_recommendation_model -- \
  --features user_history,template_effectiveness,context \
  --target completion_and_rating \
  --output models/recommendation_ranker.model

# Evaluate model
cargo run --bin evaluate_recommendation_model -- \
  --model models/recommendation_ranker.model \
  --test-data data/test_recommendations.csv
```

## Success Metrics

- ML model accuracy: >60% on completion prediction
- Collaborative recommendations acceptance: >30%
- Trajectory prediction MAE: <10 points
- Protocol completion rate: >50%
- User engagement with smart features: >20% adoption

## Related Issues

- Parent: #118 (Phase 8: Advanced Recovery Recommendation Engine)
- Depends on: #129, #130, #131, #132, #133

**Comment by jpequegn** [comment]
## Implementation Plan - Sub-Issues Created

Based on the completeness review, Phase 8.6 has been broken down into 6 trackable sub-issues:

### Phase 1: High-Value, User-Facing (4 days)
- #178 - Recovery Protocols (2 days)
- #179 - Progressive Recommendations (2 days)

### Phase 2: Enhanced UX (2 days)
- #180 - Context-Aware Features (2 days)

### Phase 3: Advanced Personalization (5-6 days)
- #181 - Collaborative Filtering (2-3 days)
- #182 - Time-Series Forecasting (3 days)

### Phase 4: ML/Advanced - Requires Training Data (3-4 days)
- #183 - ML-Based Ranking (3-4 days)

**Total Estimated Time**: 14-16 days

**Current Status**: Phase 8.6 is ~2% complete. Foundation from #129-133 is solid and ready.

**Recommendation**: Start with Phase 1 (#178, #179) for immediate user value while collecting data for ML features in Phase 4.
